### PR TITLE
Add bundling of 3rdparty apps for receipt-scanner

### LINKED ISSUE #2: Bundle binaries
We need to bundle the following binaries in the app:
- opencv3
- poppler
- tesseract

### PR DESCRIPTION
**Pull request by danschultzer**
Resolves #2 
- [x] Remove unnecessary files
  - [x] Remove all `share/man` files
  - [x] Remove all `include` files
- ~~Set `PKG_CONFIG_PATH` env var on `npm install`~~ (added readme instead)
- [x] Only build on Mac OS X systems (don't build on Circle CI)
- ~~Use Core Graphics instead of compiling libjpeg, libtiff, etc?~~ (can't be used)
- [x] Test on sandboxed/clean system to see that the binaries work as expected
- [x] Test JPEG2000 files

### CODE REVIEW

**Comment by danschultzer** [comment]
Release builds now work, but with one caveat. To build it, you need to define `PKG_CONFIG_PATH`. Run:

``` sh
PKG_CONFIG_PATH=$(pwd)/thirdparty/opencv/lib/pkgconfig npm run release
```

This should be included somewhere in the dev setup though so it happens automatically.

Current file size is 389.6 MB. Breakdown:
- `app.asar`: 78.5 MB
- `electron`: 116.9 MB
- `thirdparty`: 194.7 MB
  - `thirdparty/dependencies`: 29.8 MB
  - `thirdparty/opencv`: 53.7 MB
  - `thirdparty/opencv`: 14.8 MB
  - `thirdparty/tesseract`: 96.4 MB

**Comment by danschultzer** [comment]
Overall this PR is ready for 👀, and can be shipped. There're a few minor things that can be done, namely removing all unnecessary files. We can probably remove 20MB+.

**Comment by danschultzer** [comment]
Current file size is 332.8 MB. Breakdown:
- `app.asar`: 49.6 MB
- `thirdparty`: 174.8 MB
  - `thirdparty/dependencies`: 20.7 MB
  - `thirdparty/opencv`: 43.6 MB
  - `thirdparty/opencv`: 14.8 MB
  - `thirdparty/tesseract`: 95.6 MB

DMG: 103.5 MB
Zip: 109.9 MB

I think this is good for now. @Schultzer can you take a 👀 on this and let me know if there's anything else we could rip out of third party dependencies.

**Comment by danschultzer** [comment]
Removed a lot of unnecessary opencv libraries.

Current file size is 302.6 MB. Breakdown:
- `app.asar`: 49.6 MB
- `thirdparty`: 143.1 MB
  - `thirdparty/dependencies`: 20.7 MB
  - `thirdparty/opencv`: 12 MB
  - `thirdparty/poppler`: 14.8 MB
  - `thirdparty/tesseract`: 95.6 MB

DMG: 94.2 MB
Zip: 99.1 MB

**Comment by danschultzer** [comment]
I've added all eng related `tessdata` files now since we excluded quite a bit by just using `eng.traineddata`. Also made Travis cache much smaller by not caching `tessdata-master`.

**Comment by danschultzer** [comment]
Removed a lot of unnecessary `opencv` modules, but this causes issues upstream with `node-opencv`. I'll see if I can make a temporary fix that we can use.

```
In file included from ../src/init.cc:1:
In file included from ../src/OpenCV.h:19:
/Users/dan/blazing-bookkeeper/thirdparty/opencv/include/opencv/cv.h:66:10: fatal error: 'opencv2/video/tracking_c.h' file not found
#include "opencv2/video/tracking_c.h"
```

**Comment by danschultzer** [comment]
Was a lot of work to get [node-opencv](https://github.com/peterbraden/node-opencv) to work with available modules, but it's working now. I'll put up a PR soon to fix it upstream, but for now, we'll just use the branch to build.

**Comment by danschultzer** [comment]
The last detail in this is `libopenjpeg` and `jasper`. `Poppler` throws this warning when we compile it without `libopenjpeg`:

> Warning: Using libopenjpeg is recommended. The internal JPX decoder is unmaintained.

I'm not sure if we should support JPEG 2000 (`.jp2` and `.jpx`) or not. I would vote for disabling support completely, but we can do a quick test with PDF containing jpeg 2K images to see how it'll work or break.

The support is limited: http://caniuse.com/#feat=jpeg2000

**Comment by danschultzer** [comment]
Removed `libopenjpeg` completely. Ready for final review, and this can be merged in.

File: src/utils/thirdparty_env.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import getResourcesPath from './resources'
+
+var thirdpartyPath = getResourcesPath('thirdparty'),
+  env = {};
+
+if (process.platform == 'darwin') {
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `(process.platform == 'darwin')` this should be `(process.platform === 'darwin')`

File: scripts/build/darwin/opencv.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+then
+  cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=$THIRDPARTYDIR/opencv \
+    -D BUILD_SHARED_LIBS=ON -D BUILD_opencv_apps=OFF -D BUILD_DOCS=OFF -D BUILD_TESTS=OFF -D BUILD_PERF_TESTS=OFF -D BUILD_EXAMPLES=OFF -D BUILD_PACKAGE=OFF -D BUILD_FAT_JAVA_LIB=OFF -D BUILD_CUDA_STUBS=OFF \
+    -D WITH_IPP=ON \
+    -D WITH_OPENCL=OFF -D WITH_CUDA=OFF -D BUILD_opencv_gpu=OFF -D BUILD_opencv_gpuarithm=OFF -D BUILD_opencv_gpubgsegm=OFF -D BUILD_opencv_gpucodec=OFF -D BUILD_opencv_gpufeatures2d=OFF -D BUILD_opencv_gpufilters=OFF -D BUILD_opencv_gpuimgproc=OFF -D BUILD_opencv_gpulegacy=OFF -D BUILD_opencv_gpuoptflow=OFF -D BUILD_opencv_gpustereo=OFF -D BUILD_opencv_gpuwarping=OFF -D BUILD_opencv_shape=OFF -D BUILD_opencv_stiching=OFF -D BUILD_opencv_superres=OFF -D BUILD_opencv_objdetect=OFF \
+    -D BUILD_TIFF=OFF -D BUILD_PNG=OFF -D BUILD_JPEG=OFF -D BUILD_ZLIB=OFF -D WITH_JASPER=OFF -D WITH_OPENEXR=OFF \
+    -D BUILD_opencv_highgui=OFF -D BUILD_opencv_flann=OFF -D BUILD_opencv_ml=OFF -D BUILD_opencv_calib3d=OFF -D BUILD_opencv_videoio=OFF -D BUILD_opencv_video=OFF -D BUILD_opencv_videostab=OFF -D BUILD_opencv_java=OFF -DBUILD_opencv_python=OFF \
```
Comment (Schultzer) [inline scripts/build/darwin/opencv.sh]: `-DBUILD_opencv_python=OFF`  This should be `-D BUILD_opencv_python=OFF`

File: src/utils/thirdparty_env.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  // Set 3rd party binaries and libraries
+  env.PATH = [
+    "$PATH",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"$PATH"` should be `'$PATH'` we should use single quotation marks for strings

File: src/utils/thirdparty_env.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/poppler/bin",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/poppler/bin"` should be `'/poppler/bin'`

File: src/utils/thirdparty_env.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/tesseract/bin"
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/tesseract/bin"` should be `/tesseract/bin`

File: src/utils/thirdparty_env.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  ].join(':');
+
+  env.DYLD_LIBRARY_PATH = [
+    "$DYLD_LIBRARY_PATH",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"$DYLD_LIBRARY_PATH"` should be `'$DYLD_LIBRARY_PATH'`

File: src/utils/thirdparty_env.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    "/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/", // Core Graphics package
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/"` should be `'/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/'`

File: src/utils/thirdparty_env.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/dependencies/lib",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/dependencies/lib"` should be `'/dependencies/lib'`

File: src/utils/thirdparty_env.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/poppler/lib",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/poppler/lib"` should be `'/poppler/lib'`

File: src/utils/thirdparty_env.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/opencv/lib",
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/opencv/lib"` should be `'/opencv/lib'`

File: src/utils/thirdparty_env.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    thirdpartyPath + "/tesseract/lib"
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/tesseract/lib"` should be `'/tesseract/lib'`

File: src/utils/thirdparty_env.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  ].join(':');
+
+  env.TESSDATA_PREFIX = thirdpartyPath + "/tesseract/share/tessdata";
```
Comment (Schultzer) [inline src/utils/thirdparty_env.js]: `"/tesseract/share/tessdata"` should be `'/tesseract/share/tessdata'`

File: tasks/build_app.js [hunk 13]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+gulp.task('dependencies', function () {
+    var rootDir = jetpack.cwd('./').path(),
+        platform = process.platform,
+        buildScript = jetpack.cwd('./scripts').path('build/' + platform + '/build.sh'),
+        buildDir = rootDir + '/build',
+        buildDestDir = rootDir + '/thirdparty';
+
+      return new Promise(function(resolve, reject) {
+          if (!jetpack.exists(buildScript)) {
+            gutil.log("No build file for", gutil.colors.cyan("'" + platform + "'"));
```
Comment (Schultzer) [inline tasks/build_app.js]: `gutil.log("No build file for", gutil.colors.cyan("'" + platform + "'"));` should be `gutil.log('No build file for', "'" +  gutil.colors.cyan(platform) + "'");` we don't want the qutation marks to be cyan but white.

File: tasks/build_app.js [hunk 14]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+            return resolve();
+          }
+
+          gutil.log("Going to build for", gutil.colors.cyan("'" + platform + "'"));
```
Comment (Schultzer) [inline tasks/build_app.js]: `gutil.log("Going to build for", gutil.colors.cyan("'" + platform + "'"));` should be `gutil.log('Going to build for', "'" +  gutil.colors.cyan(platform) + "'");`

File: tasks/build_app.js [hunk 15]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+
+          var buildCommand = [
+            "BUILDDIR=" + buildDir,
```
Comment (Schultzer) [inline tasks/build_app.js]: `"BUILDDIR="` should be `'BUILDDIR='`

File: tasks/build_app.js [hunk 16]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+            "THIRDPARTYDIR=" + buildDestDir,
```
Comment (Schultzer) [inline tasks/build_app.js]: `"THIRDPARTYDIR="` should be `'THIRDPARTYDIR='`

File: tasks/build_app.js [hunk 17]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+            'sh ' + buildScript,
+          ].join(" ");
```
Comment (Schultzer) [inline tasks/build_app.js]: `join(" ");` should be `join(' ');`

File: tasks/build_app.js [hunk 18]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+          var child = exec(buildCommand, {
+              maxBuffer: 1024 * 1024 * 10
+            },
+            function(error, stdout, stderr) {
+              if (!error) {
+                  gutil.log("Build succesful");
```
Comment (Schultzer) [inline tasks/build_app.js]: `gutil.log("Build succesful");` should be `gutil.log('Build succesful');`

File: tasks/build_app.js [hunk 19]
```diff
@@ -4,15 +4,56 @@ var gulp = require('gulp');
...
+                  resolve();
+              } else {
+                  gutil.log("Build failed");
```
Comment (Schultzer) [inline tasks/build_app.js]: `gutil.log("Build failed");` should be `gutil.log('Build failed');`

**Review by Schultzer — CHANGES_REQUESTED** [review]
We should follow #21 with a uniform syntax and use single quotation marks and only use double quotation marks when escaping single quotation in out JS code base.

**Review by Schultzer — APPROVED** [review]

**Pull request merged**